coreneuron/gpu/nrn_acc_manager.cpp

```diff
@@ -31,15 +32,161 @@
 #include <cuda_runtime_api.h>
 #endif
 
+#if __has_include(<cxxabi.h>)
+#define USE_CXXABI
+#include <cxxabi.h>
+#include <memory>
+#include <string>
+#endif
+
+#ifdef CORENEURON_ENABLE_PRESENT_TABLE
+#include <cassert>
+#include <cstddef>
+#include <iostream>
+#include <map>
+#include <shared_mutex>
+namespace {
+std::map<std::byte const*, std::pair<std::size_t, std::byte*>> present_table;
+std::shared_mutex present_table_mutex;
+}  // namespace
+#endif
+
+namespace {
+/** @brief Try to demangle a type name, return the mangled name on failure.
+ */
+std::string cxx_demangle(const char* mangled) {
+#ifdef USE_CXXABI
+    int status{};
+    // Note that the third argument to abi::__cxa_demangle returns the length of
+    // the allocated buffer, which may be larger than strlen(demangled) + 1.
+    std::unique_ptr<char, decltype(free)*> demangled{
+        abi::__cxa_demangle(mangled, nullptr, nullptr, &status), free};
+    return status ? mangled : demangled.get();
+#else
+    return mangled;
+#endif
+}
+bool cnrn_target_debug_output_enabled() {
+    const char* env = std::getenv("CORENEURON_GPU_DEBUG");
+    if (!env) {
+        return false;
+    }
+    std::string env_s{env};
+    if (env_s == "1") {
+        return true;
+    } else if (env_s == "0") {
+        return false;
+    } else {
+        throw std::runtime_error("CORENEURON_GPU_DEBUG must be set to 0 or 1 (got " + env_s + ")");
```

Comment:
As an example, how is the output of `CORENEURON_GPU_DEBUG=1`?

Reply:
```
/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/coreneuron/gpu/nrn_acc_manager.cpp:658: cnrn_target_copyin<int>(0xfd2180, 10 * 4 = 40) -> 0x7fff97b19c00
/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/coreneuron/gpu/nrn_acc_manager.cpp:659: cnrn_target_memcpy_to_device<int*>(0x7fff97b19428, 0x7fffffff5f30, 1 * 8 = 8)
/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/coreneuron/gpu/nrn_acc_manager.cpp:727: cnrn_target_copyin<unsigned long>(0, 0 * 8 = 0) -> 0
/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/coreneuron/gpu/nrn_acc_manager.cpp:729: cnrn_target_memcpy_to_device<unsigned long*>(0x7fff97afa980, 0x7fffffff5aa0, 1 * 8 = 8)
/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/coreneuron/gpu/nrn_acc_manager.cpp:733: cnrn_target_copyin<unsigned long>(0, 0 * 8 = 0) -> 0
/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/coreneuron/gpu/nrn_acc_manager.cpp:734: cnrn_target_memcpy_to_device<unsigned long*>(0x7fff97afa990, 0x7fffffff5a48, 1 * 8 = 8)
x86_64/corenrn/mod2c/passive.cpp:176: cnrn_target_device_ptr<double>(0xfb2ec0) -> 0x7fff97b00300
x86_64/corenrn/mod2c/passive.cpp:177: cnrn_target_device_ptr<double>(0xfb34c0) -> 0x7fff97b00900
x86_64/corenrn/mod2c/passive.cpp:178: cnrn_target_device_ptr<double>(0xfb3ac0) -> 0x7fff97b00f00
x86_64/corenrn/mod2c/passive.cpp:179: cnrn_target_device_ptr<double>(0xfb40c0) -> 0x7fff97b01500
x86_64/corenrn/mod2c/passive.cpp:180: cnrn_target_device_ptr<double>(0xfb46c0) -> 0x7fff97b01b00
x86_64/corenrn/mod2c/passive.cpp:239: cnrn_target_device_ptr<coreneuron::pas_Instance>(0xd94830) -> 0x7fff97b05200
x86_64/corenrn/mod2c/passive.cpp:239: cnrn_target_memcpy_to_device<coreneuron::pas_Instance>(0x7fff97b05200, 0xd94830, 1 * 56 = 56)
```
(and so on)

no doubt it could be refined more; with our private present table implementation we might also be able to provide a more human-readable dump of it...in the future...